rasa/shared/core/trackers.py

```diff
@@ -446,15 +461,53 @@ def applied_events(self) -> List[Event]:
                 self._undo_till_previous_loop_execution(
                     event.action_name, applied_events
                 )
+            elif isinstance(event, UserUttered):
+                # update event's featurization based on the future event
+                use_text_for_featurization = self._define_user_featurization(
+                    events_as_list[i + 1 :]
+                )
+                if event.use_text_for_featurization is None:
+                    event.use_text_for_featurization = use_text_for_featurization
+                # update event's entities based on the future event
+                entities = self._define_user_entities(events_as_list[i + 1 :])
```

Comment:
@Ghostvv Can we please refactor this to use `Event.apply_to` instead?

Reply:
can we do it afterwards?

Reply:
It would be a small change. One question though: Why do we need to update the entities? Which entities wouldn't be part of `UserUttered` but part of `DefinePrevUserUtteredEntities`?

Reply:
the ones the ted predicted

Reply:
Event.apply_to should iterate over the list of events and update previous `UserUttered`?

Reply:
Mhm, the entities make it problematic as they would be stored in the database in case `UserUttered` would saved to the database after `appy_to` was called. Pretty sure this problem also exists in the current solution as `applied_events` doesn't copy events which means we only move around references to the event which means by updating an event in `applied_events` it's also changed in `tracker.events`. The solution. could be to define a separate field `entities_by_policies` in `UserUttered` so we can deal with it separately. 

In theory `Event.apply_to` could directly check the last event and update it with the information. It's in my opinion the cleaner solution as `apply_to` is called as soon as the tracker is constructed.

Reply:
we cannot do separate field, because from the point of view of the user it doesn't matter who extracted the entities

> In theory Event.apply_to could directly check the last event and update it with the information. It's in my opinion the cleaner solution as apply_to is called as soon as the tracker is constructed.

you just said above that this wouldn't work

> Pretty sure this problem also exists in the current solution

but `applied_events` will reapply the necessary changes every time, so if it was not persisted, it'll update the events

Reply:
Yep, `apply_to` will do that, too :-) The problem is if it does that too early, then the policy entities are saved as part of the `UserUttered` event (which we don't want)

> we cannot do separate field, because from the point of view of the user it doesn't matter who extracted the entities

Yeah of course. We don't have to separate it for things outside of the event but internal in the event so we can distinguish where information came from when storing the event (which uses `as_dict`)

Reply:
> you just said above that this wouldn't work

It works. The entities might be a problem but as said I'm sure that they're **already** a problem

Reply:
```

def test_tobi():
    tracker = DialogueStateTracker.from_events(
        "Vova",
        evts=[
            ActionExecuted(ACTION_LISTEN_NAME),
            UserUttered("hi", intent={"name": "greet"}),
            DefinePrevUserUtteredFeaturization(True),
            DefinePrevUserUtteredEntities(
                entities=[{"entity": "entity1", "value": "value1"}]
            ),
        ],
    )

    tracker.applied_events()

    user = list(tracker.events)[1]
    assert isinstance(user, UserUttered)

    # Fails
    assert not user.entities
```

Reply:
> then the policy entities are saved as part of the UserUttered event
why?

Reply:
> assert not user.entities
I'm confused. that's the whole point, we want to update entities in previous `UserUttered` event

Reply:
created the PR: https://github.com/RasaHQ/rasa/pull/7503

Reply:
> I'm confused. that's the whole point, we want to update entities in previous UserUttered event

Yes, you're absolutely right. The problem is the persistence. Updating the `entities` in `UserUttered` is great for runtime. However, we will also persist the updated entities as part of `UserUttered`. When you then load the tracker from the tracker store again, you'll obtain a changed history.

Reply:
do you mean that featurization will change? that should be fine, we handle this case in `as_sub_state` where we remove all the entities as core input features if user uttered should be featurized as text

Reply:
@tmbo I'm sorry to pull you in here but as I suspect that you implemented events / tracker store in the first place your opinion on this would be great 🙌 

**Background**
Policies now add events (`DefinePrevUserUtteredEntities`, `DefinePrevUserUtteredFeaturization`) to the tracker. These events add additional information to the previous `UserUttered` event. When these events are applied to the tracker, they will update `UserUttered` on the fly with the additional properties so that this information is available when the tracker is featurized for the ML components.

**The issue**
From a perspective of the ML components updating the `UserUttered` events is fine in my opinion. The problem is that `DefinePrevUserUtteredEntities` updates the `entities` of `UserUttered` before the `UserUttered` event is persisted to the `TrackerStore`. This means that when we retrieve and de-serialize such a tracker from the `TrackerStore`, it will look like as if the `entities` were part of the `UserUttered` event from the beginning (and not added by the policies). 

This means that the immutability property of events is broken as the history in the `TrackerStore` does not reflect the past 1:1 anymore. This might seem like a small hack or even convenience at this point, but losing access to the source of truth  seems like opening the box of the pandora. 

**Other Notes**
* Vova is changing the logic slightly in [this PR](https://github.com/RasaHQ/rasa/pull/7503) which moves the logic from `applied_events` to `apply_to` of the events.
* There is also some [more discussions](https://github.com/RasaHQ/rasa/pull/7503#discussion_r540180056) in the other PR about the same issue.

Reply:
The box is already opened, I patched it through core featurizers, but we should definitely redesign the system

Reply:
I am not sure I have enough context, but I'll try my best:

We can't change past events. 

The events should be viewed like atomic changes applied to a state and are streamed to outside consumers (e.g. Rasa X). Starting with an initial state and applying all events in the sequence should always get you to the same state. (side notes: they are commonly phrased in the past tense symbolizing that something happened `UserUttered` rather than using a command `Define....` (e.g. `EntityDetected` sounds a lot simpler)).

What does `DefinePrevUserUtteredFeaturization ` do?

> does not reflect the past 1:1 anymore. This might seem like a small hack or even convenience at this point, but losing access to the source of truth seems like opening the box of the pandora.

I think I am lacking more detail, but this doesn't sound like something we can do while making sure external event consumers are able to reproduce the state